File: native/iosApp/ReportediOS/Sources/UdfStore.swift
import Combine
import Foundation

/// The single state/action boundary used by iOS features.
///
/// Views render `state` and send user or system events through `onAction`.
/// Implementations own all feature-state mutation and side effects.
@MainActor
protocol UdfStore: ObservableObject {
    associatedtype State
    associatedtype Action
    associatedtype ActionResult

    var state: State { get }
    func onAction(_ action: Action) -> ActionResult
}

File: native/androidApp/src/main/java/com/reported/nativeandroid/app/UdfStore.kt
package com.reported.nativeandroid.app

import kotlinx.coroutines.flow.StateFlow

/**
 * The single state/action boundary used by Android features.
 *
 * Views render [state] and send user or system events through [onAction]. State
 * mutations and side effects belong in the implementing store/ViewModel.
 */
interface UdfStore<State : Any, Action : Any> {
    val state: StateFlow<State>

    fun onAction(action: Action)
}
